<?xml version="1.0" encoding="UTF-8"?>
<launch>
  <arg name="ns" default=""/>
  <arg name="image_compressed_topic"/>
  <arg name="camera_info_topic"/>
  <arg name="filtered_detections_topic" default="apriltag/filtered/detections_array"/>
  <arg name="apriltag_detector_families"/>

  <group>
    <push-ros-namespace namespace="$(var ns)"/>

    <node pkg="autoware_image_transport_decompressor" exec="image_transport_decompressor_node" name="decompressor" output="screen">
      <remap from="decompressor/input/compressed_image" to="$(var image_compressed_topic)"/>
      <remap from="decompressor/output/raw_image" to="raw_image"/>

      <param name="encoding" value="default"/>
    </node>

    <include file="$(find-pkg-share tag_based_sfm_calibrator)/launch/apriltag_detector.launch.py">
      <arg name="image_topic" value="raw_image"/>
      <arg name="camera_info_topic" value="$(var camera_info_topic)"/>
      <arg name="families" value="$(var apriltag_detector_families)"/>
    </include>

    <node pkg="tier4_tag_utils" exec="apriltag_filter_node" name="apriltag_filter" output="screen">
      <remap from="camera_info" to="$(var camera_info_topic)"/>
      <remap from="apriltag/filtered/detection_array" to="$(var filtered_detections_topic)"/>
      <!-- filtering parameters -->
      <param name="min_tag_size" value="0.6"/>
      <!--0.6 = 0.8 * 0.75 the apriltag size, not the lidartag-->
      <param name="max_tag_distance" value="30.0"/>
      <param name="max_allowed_homography_error" value="0.5"/>
      <param name="min_margin" value="30.0"/>
      <param name="max_hamming_error" value="0"/>
      <param name="max_no_observation_time" value="3.0"/>
      <param name="new_hypothesis_distance" value="1.5"/>
      <param name="tag_families" value="$(var apriltag_detector_families)"/>
      <!-- the sizes of tag_ids -->
      <param name="max_convergence_translation" value="2.0"/>
      <param name="new_hypothesis_translation" value="20.0"/>

      <param name="measurement_noise_translation" value="0.2"/>
      <param name="process_noise_translation" value="0.02"/>
    </node>
  </group>
</launch>
